### PR TITLE
Issues/dharmik 8440 reset password field

### PR DESCRIPTION
**Pull request by dharmik**
### Fixed Issues

$ https://github.com/Expensify/App/issues/8440

### Tests

Expected changes:
https://github.com/Expensify/App/issues/8440#issuecomment-1103186938

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps

1. Open the New Expensify App
2. Sign into any valid account
3. Tap on the "Profile" icon on the top left
4. Tap on "Security"
5. Tap on "Change Password""
6. Enter an incorrect password in the "Current Password" field
7. Enter a correct Password that follows the requirement in "New Password" field
8. Tap on green "Save" button
9. Steps to verify https://github.com/Expensify/App/issues/8440#issuecomment-1103186938

- [x] Verify that no errors appear in the JS console

### Photo

| Platform  | Image |
| ------------- | ------------- |
| Web  | https://user-images.githubusercontent.com/1510175/167183342-757e2771-9708-4b29-86ba-1cacd6f73cf4.png |
| iOS  |   https://user-images.githubusercontent.com/1510175/164319032-56f95161-e1d4-4263-a08b-26979ed1a0f8.png|
| Mobile Web  |   https://user-images.githubusercontent.com/1510175/167180979-3022f8e5-b545-4013-8597-f587a0bd341b.png|
| Desktop  |   https://user-images.githubusercontent.com/1510175/167181083-d41fadf1-0b2c-4dc9-a9ff-2554405e48cf.png|
| Android  |   https://user-images.githubusercontent.com/1510175/167182403-364e37b7-70f8-4d85-ac03-f949f1124d01.png|

### CODE REVIEW

**Review by parasharrajat — DISMISSED** [review]
LGTM. 

cc: @NikkiWines 

:ribbon: :eyes: :ribbon:  C+ reviewed

File: src/libs/actions/User.js [hunk 1]
```diff
@@ -53,7 +53,8 @@ function changePasswordAndNavigate(oldPassword, password) {
         .then((response) => {
             if (response.jsonCode !== 200) {
                 const error = lodashGet(response, 'message', 'Unable to change password. Please try again.');
-                Onyx.merge(ONYXKEYS.ACCOUNT, {error});
+                const errorCode = lodashGet(response, 'jsonCode', 0);
+                Onyx.merge(ONYXKEYS.ACCOUNT, {error, errorInCurrentPassword: errorCode === 401, errorToNotAllowPasswordChange: errorCode === 666});
```
Comment (NikkiWines) [inline src/libs/actions/User.js:57]: I know this is a bit against what we discussed in the issue but what if instead of `errorInCurrentPassword` and `errorToNotAllowPasswordChange` we used something like `errorType`, where `errorType` can be `current`, `new`, or `generic`. And then we could follow the convention [here](https://github.com/Expensify/App/blob/42a9518275a936673b73e6b2272a90faaa3785b8/src/libs/actions/Session/index.js#L456-L461) and do something like: 

```Javascript 
            if (errorCode.jsonCode === 401) {
                Onyx.merge(ONYXKEYS.SESSION, {errorType: 'current'});
            }
            if (errorCode.jsonCode === 666) {
                Onyx.merge(ONYXKEYS.ACCOUNT, {errorType: 'generic'});
            }
```

File: src/pages/settings/PasswordPage.js [hunk 2]
```diff
@@ -180,7 +181,7 @@ class PasswordPage extends Component {
                                             value={this.state.currentPassword}
                                             onChangeText={text => this.clearErrorAndSetValue('currentPassword', text)}
                                             returnKeyType="done"
-                                            hasError={this.state.errors.currentPassword}
+                                            hasError={this.state.errors.currentPassword || this.props.account.errorInCurrentPassword}
                                             errorText={this.getErrorText('currentPassword')}
```
Comment (NikkiWines) [inline src/pages/settings/PasswordPage.js:185]: `errorText` should be removed if we're following the new styling conventioned outlined in [this comment](https://github.com/Expensify/App/issues/8440#issuecomment-1103186938)

Otherwise we still get the following display 

image

**Review by NikkiWines — CHANGES_REQUESTED** [review]
A couple of suggestions. 

Also, I'm not sure if I misunderstood something in the proposal stage but did we agree to update the UI to [the following style](https://github.com/Expensify/App/issues/8440#issuecomment-1103186938)? 

Lastly, some small spacing issues - the spacing differs for errors from the API vs. pre-submission validation errors

image

vs. 

image

**Comment by dharmik** [comment]
@NikkiWines As we are using `<TextInput/>` and we display errorText  inside that component so as per consistent with other layout we display that same.
Example at other place for : **pre-submission validation** as well as for API error.
Simulator Screen Shot - iPhone 13 - 2022-05-18 at 00 03 51

File: src/libs/actions/User.js (hunk 1 above)
Comment (dharmik) [reply src/libs/actions/User.js:57]: Resolved.

**Comment by NikkiWines** [comment]
> As we are using <TextInput/> and we display errorText inside that component so as per consistent with other layout we display that same.

Yeah, in this case, I'm suggesting we remove the error messages from `<TextInput/>` elements and instead display the messages in the `<Text style={styles.formError}>` element (and adjust the spacing as necessary), but retain the red outlining around the specific fields when appropriate. 

It's a good point that the proposed UI change is a departure from existing patterns. In this case, the design team did sign off on the [proposed change](https://github.com/Expensify/App/issues/8440#issuecomment-1103186938) (One of our designers, @shawnborton, thumbs up'd, and another design team member commented their approval) so I think we're all set to implement the suggested UI update.

**Comment by dharmik** [comment]
@NikkiWines So we have to change in all other places so can we create separate task for it? Those changes take some time and additional efforts to check in all platforms. Just a suggestion.

**Comment by shawnborton** [comment]
Regarding the error spacing (indented from the left edge or not)... I think we've gone back and forth there. Do we have a standard in the app somewhere? cc @roryabraham as I think we spoke about this a long time ago.

**Comment by NikkiWines** [comment]
> @NikkiWines So we have to change in all other places so can we create separate task for it? Those changes take some time and additional efforts to check in all platforms. Just a suggestion.

Sorry @dharmik, I'm not sure I understand. I'm suggesting this change only for [src/pages/settings/PasswordPage.js](https://github.com/Expensify/App/pull/8905/files/396d0576d730f5a310afbb7b3c4afda8dd2021bb#diff-b4474bebdf5744f78c9eded74e6bd3e4d2723040f750e573f4ef4bddfe00481b), not for all components that use `<TextInput/>` with error messages. It shouldn't be too extensive to update and test.

**Comment by roryabraham** [comment]
@shawnborton I don't know if we standardized, but I was always on team "indent from the left edge"

**Comment by shawnborton** [comment]
Okay, then I say we just use whatever the components are currently doing and then we can follow up in a separate issue and make sure everything is aligned to whatever we end up deciding.

**Comment by dharmik** [comment]
@shawnborton Thanks for information so wi will cover that UI related change in a separate issue.

@NikkiWines Logic which you suggest those change are committed. Once you have some please review it.

File: src/pages/settings/PasswordPage.js [hunk 3]
```diff
@@ -240,6 +241,7 @@ export default compose(
     withOnyx({
         account: {
             key: ONYXKEYS.ACCOUNT,
+            errorType: '',
```
Comment (parasharrajat) [inline src/pages/settings/PasswordPage.js:244]:
```suggestion
```

File: src/pages/settings/PasswordPage.js [hunk 4]
```diff
@@ -192,7 +193,7 @@ class PasswordPage extends Component {
                                             autoCompleteType="password"
                                             textContentType="password"
                                             value={this.state.newPassword}
-                                            hasError={this.state.errors.newPassword || this.state.errors.newPasswordSameAsOld}
```
Comment (parasharrajat) [inline src/pages/settings/PasswordPage.js:195]: Why did we remove the error check for `newPasswordSameAsOld`?

File: src/libs/actions/User.js [hunk 5]
```diff
@@ -53,7 +53,13 @@ function changePasswordAndNavigate(oldPassword, password) {
         .then((response) => {
             if (response.jsonCode !== 200) {
                 const error = lodashGet(response, 'message', 'Unable to change password. Please try again.');
-                Onyx.merge(ONYXKEYS.ACCOUNT, {error});
+                const errorCode = lodashGet(response, 'jsonCode', 0);
+                if (errorCode === 401) {
+                    Onyx.merge(ONYXKEYS.ACCOUNT, {error, errorType: 'current'});
+                }
+                if (errorCode === 666) {
+                    Onyx.merge(ONYXKEYS.ACCOUNT, {error, errorType: 'generic'});
```
Comment (parasharrajat) [inline src/libs/actions/User.js:61]: Move the current and generic to CONST.js

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/settings/PasswordPage.js (hunk 4 above)
Comment (dharmik) [reply src/pages/settings/PasswordPage.js:195]: Not we check as a
`hasError={this.state.errors.newPassword || this.props.account.errorType === 'generic'}`

if new password is same as old password at that time: errorCode === 666

File: src/libs/actions/User.js (hunk 5 above)
Comment (dharmik) [reply src/libs/actions/User.js]: Applied changes.

File: src/pages/settings/PasswordPage.js (hunk 4 above)
Comment (parasharrajat) [reply src/pages/settings/PasswordPage.js:195]: But this validation should be done on the front end. We should not delegate the validations to the backend when they can be done on the frontend.

Comment (dharmik) [reply src/pages/settings/PasswordPage.js]: @parasharrajat I apology to create the confusion. I accidentally miss that. 

PR updated.

File: src/CONST.js [hunk 6]
```diff
@@ -377,7 +377,11 @@ const CONST = {
     PASSWORD_PAGE: {
         ERROR: {
             ALREADY_VALIDATED: 'Account already validated',
-            VALIDATE_CODE_FAILED: 'Validate code failed',
+            VALIDATE_CODE_FAILED: '',
```
Comment (parasharrajat) [inline src/CONST.js:380]: Why this change?

File: src/libs/actions/User.js [hunk 7]
```diff
@@ -53,7 +53,13 @@ function changePasswordAndNavigate(oldPassword, password) {
         .then((response) => {
             if (response.jsonCode !== 200) {
                 const error = lodashGet(response, 'message', 'Unable to change password. Please try again.');
```
Comment (parasharrajat) [inline src/libs/actions/User.js:55]: Please translate this error. Move this to translation files and then use it here. You can get the Spanish translation on slack. Thanks.

Comment (dharmik) [reply src/libs/actions/User.js:55]: @parasharrajat I request in slack again today is there any person you know so I will tag that developer. I can translate from google but some time the translation are not correct.
Any suggestion please

Comment (parasharrajat) [reply src/libs/actions/User.js:55]: Let's wait for Monday.

**Comment by parasharrajat** [comment]
Testing...

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Chnages looks fine to me but UI is not aligned with https://github.com/Expensify/App/issues/8440#issuecomment-1103186938. 

- Gap above the error message is larger.

File: src/pages/settings/PasswordPage.js (hunk 2 above)
Comment (parasharrajat) [reply src/pages/settings/PasswordPage.js:185]: I feel like this is fine as it is only showing the required field error.

**Comment by dharmik** [comment]
> Chnages looks fine to me but UI is not aligned with [#8440 (comment)](https://github.com/Expensify/App/issues/8440#issuecomment-1103186938).
> 
> * Gap above the error message is larger.
It' global so it came from <TextInput> and as we discuss we have to take it on separate task.
https://github.com/Expensify/App/pull/8905#issuecomment-1129211531

**Comment by parasharrajat** [comment]
Nikkiwines asked to adjust the spacing [here](https://github.com/Expensify/App/pull/8905#issuecomment-1129288886). 

I feel like adjusting the spacing should not be worth of a new issue. This change is only specific to this page and other page are following different Form guidelines.

**Comment by dharmik** [comment]
@parasharrajat I hope you ready this one too
https://github.com/Expensify/App/pull/8905#issuecomment-1131420168

**Comment by parasharrajat** [comment]
That is about _I don't know if we standardized, but I was always on team "indent from the left edge_ which is not what I pointed out on the issue.

File: src/pages/settings/PasswordPage.js (hunk 2 above)
Comment (NikkiWines) [reply src/pages/settings/PasswordPage.js:185]: Hmm, but then we'll have two different behaviors for errors for the `currentPassword`

Screen Shot 2022-06-09 at 14 44 44
Screen Shot 2022-06-09 at 14 44 12

I think it's fairly simple to just always show the errors in place of the `passwordPage.newPasswordPrompt` below the `New Password` text box. That way we can have consistent spacing and placing of error messages and it aligns with what our design team agreed on.

**Comment by NikkiWines** [comment]
Sorry for the delay catching up here. 

@shawnborton did mention making a new issue for updating the alignment to use the left edge
> Okay, then I say we just use whatever the components are currently doing and then we can follow up in a separate issue and make sure everything is aligned to whatever we end up deciding.

However, that doesn't mean we should merge this with inconsistent spacing for error messages. The inconsistency is why I'm suggesting we have all error messages display in a separate `<Text/>` element instead of using `<TextInputs/>` `errorText` value for each text box.

**Comment by shawnborton** [comment]
I agree, let's avoid merging this with the mismatched spacing.

**Comment by dharmik** [comment]
So what's the final conclusion?
For this ticket we have to remove Error from <TextInputs/> and use <Text/> for Password change page?
@NikkiWines @shawnborton

**Comment by NikkiWines** [comment]
I personally think we should implement what I suggested [here](https://github.com/Expensify/App/pull/8905#discussion_r893848656) because it solves the original issue of the error message not being visible, maintains consistency for spacing and where error messages are being displayed, and allows us to outline the appropriate `<TextInput/>` when applicable. 

We can do this only for `PasswordPage.js` we don't have to make any changes to the other components that use `<TextInput/>`.

**Comment by NikkiWines** [comment]
Closing as the main issue for this has been closed and payment has been issued.
